README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
killedbypolice
==============

<!--
[![Last-changedate](https://img.shields.io/badge/last%20change-2018--01--09-brightgreen.svg)](https://github.com/benmarwick/killedbypolice/commits/master)  
[![Travis build status](https://travis-ci.org/benmarwick/killedbypolice.svg?branch=master)](https://travis-ci.org/benmarwick/killedbypolice)   [![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://img.shields.io/badge/lifecycle-experimental-orange.svg)
-->
The goal of killedbypolice is to make readily available the data collected by <http://killedbypolice.net/> for exploration, visualisation, and analysis.

We don't know much about who collects the data for <http://killedbypolice.net/>, or what their methods are. [FiveThirtyEight](https://fivethirtyeight.com/features/another-much-higher-count-of-police-homicides/) reported that he was 'an instructor on nonviolent physical-intervention techniques and that he prefers to remain anonymous'.

This is an important data set because the 'US government has no comprehensive record of the number of people killed by law enforcement.' ([*The Guardian*, 1 June 2015](https://www.theguardian.com/us-news/ng-interactive/2015/jun/01/about-the-counted)). The killedbypolice project is one of a few non-government projects that continuously collect data on police killings (see [Related work](#related-work) below).

Installation
------------

You can install killedbypolice from github with:

``` r
# install.packages("devtools")
devtools::install_github("benmarwick/killedbypolice")
```

How to use
----------

This is a basic example which shows you how to access the data in this package:

``` r
# load the library
library(killedbypolice)

# load the data frame, ready to work with 
data("kbp2013_2017")
```

Here are some ways we can take a look at the structure of the data:

``` r
# inspect the data
suppressPackageStartupMessages(library(dplyr))
glimpse(kbp2013_2017)
#> Observations: 5,578
#> Variables: 13
#> $ state          <chr> "WA", "CA", "NC", "PA", "CA", "CO", "CA", "LA",...
#> $ name           <chr> "James M. Eshelman", "Dontae Hayes", "Ricky Jun...
#> $ age            <dbl> 54, 20, 34, 43, 31, 43, 45, 22, 30, 51, 37, 40,...
#> $ gender         <chr> "M", "M", "M", "M", "F", "M", "M", "M", "M", "M...
#> $ race_ethnicity <chr> NA, "B", "B", "B", "W", "W", "O", NA, "W", "W",...
#> $ method_1       <chr> "G", "G", "G", "G", "G", "G", "G", "G", "G", "G...
#> $ method_2       <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,...
#> $ method_3       <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,...
#> $ method_4       <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,...
#> $ method_5       <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA,...
#> $ event_year     <dbl> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013,...
#> $ event_month    <dbl> 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12, 12,...
#> $ event_day      <int> 31, 31, 31, 30, 30, 29, 29, 29, 29, 29, 28, 28,...

head(kbp2013_2017)
#> # A tibble: 6 x 13
#>   state name     age gend~ race~ meth~ meth~ meth~ meth~ meth~ even~ even~
#>   <chr> <chr>  <dbl> <chr> <chr> <chr> <chr> <chr> <chr> <chr> <dbl> <dbl>
#> 1 WA    James~  54.0 M     <NA>  G     <NA>  <NA>  <NA>  <NA>   2013  12.0
#> 2 CA    Donta~  20.0 M     B     G     <NA>  <NA>  <NA>  <NA>   2013  12.0
#> 3 NC    Ricky~  34.0 M     B     G     <NA>  <NA>  <NA>  <NA>   2013  12.0
#> 4 PA    Willi~  43.0 M     B     G     <NA>  <NA>  <NA>  <NA>   2013  12.0
#> 5 CA    Betha~  31.0 F     W     G     <NA>  <NA>  <NA>  <NA>   2013  12.0
#> 6 CO    Chris~  43.0 M     W     G     <NA>  <NA>  <NA>  <NA>   2013  12.0
#> # ... with 1 more variable: event_day <int>
```

Some explorations
-----------------

Here are some explorations of the data:

The most common age to be killed by police is in the late twenties and early thirties, and this has not changed much over time.

``` r
library(ggplot2)
library(ggridges)

kbp2013_2017 %>% 
  filter(gender %in% c("F", "M", "T")) %>% 
  filter(!is.na(event_year)) %>% 
  ggplot(aes(x = age,
             y = factor(event_year),
             fill = gender)) +
  geom_density_ridges(alpha = 0.5, 
                      scale = 0.9)  +
  theme_ridges(font_size = 10) +
  scale_x_continuous(breaks = seq(0, 100, 10),
                     labels = seq(0, 100, 10)) +
  xlab("Age at death (years)") +
  ylab("Year") +
  theme(axis.title = element_text(size = 14))
```

![](readme-figs/unnamed-chunk-2-1.png)

Of the three ethnic groups that make up most of the deaths, Black and Latino people tend to be younger than White people when they are killed by police.

``` r
kbp2013_2017 %>% 
  filter(race_ethnicity %in% c("B", "W", "L")) %>% 
  filter(!is.na(event_year)) %>% 
  ggplot(aes(x = age,
             y = factor(event_year),
             fill = race_ethnicity)) +
  geom_density_ridges(alpha = 0.6, 
                      scale = 0.9)  +
  theme_ridges(font_size = 10) +
  scale_x_continuous(breaks = seq(0, 100, 10),
                     labels = seq(0, 100, 10)) +
  xlab("Age at death (years)") +
  ylab("Year") +
  theme(axis.title = element_text(size = 14))
```

![](readme-figs/unnamed-chunk-3-1.png)

By far the most common way that people are killed by police is with a gun. Deaths by vehicle involve women more often than men.

``` r
kbp2013_2017 %>% 
  filter(!is.na(event_year)) %>% 
  filter(method_1 != "NA") %>% 
  filter(gender %in% c("M", "F", "T")) %>% 
  group_by(event_year, 
           gender,
           method_1) %>% 
  tally() %>% 
  mutate(perc = n / sum(n) * 100)  %>% 
  ggplot(aes(method_1,
             perc,
             fill = gender)) +
  geom_col() +
  facet_grid(gender~event_year) +
  theme_minimal(base_size = 14) +
  xlab("Method of killing") +
  ylab("Percentage of all\npeople killed by police\nby gender") 
```

![](readme-figs/unnamed-chunk-4-1.png)

In 2016, the state with the largest number of people killed by police was California.

``` r
library(statebins) # using GitHub version
library(viridis)

# we need to convert state abbreviations to state names for the statebins function
state_abb <- data_frame(state_name = state.name,
                        state_abb = state.abb)

# we need to add the state popluations so we can get a proportion of people in each state
# we got this from https://www2.census.gov/programs-surveys/popest/tables/2010-2016/state/totals/nst-est2016-01.xlsx
state_populations <- readr::read_csv("data-raw/nst-est2016-01.csv")

# clean it a little
state_populations <-  
  state_populations %>% 
  mutate(state_name = gsub("\\.", "", X__1)) %>%
  left_join(state_abb)

# compute deaths by state and as deaths per 1000 people in each state
by_state <- kbp2013_2017 %>% 
  filter(event_year == 2016) %>% 
  group_by(state) %>% 
  tally() %>% 
  left_join(state_abb, by = c('state' = 'state_abb')) %>% 
  filter(!is.na(state_name)) %>% 
  left_join(state_populations) %>% 
  mutate(per_n_people = (n / `2016`) * 1000000)

# plot 'statebin' style map
ggplot(by_state, 
       aes(state = state_name, 
           fill = n)) +
  geom_statebins() +
  coord_equal() +
  scale_fill_viridis() +
  theme_statebins() +
  labs(title = "Total number of people killed by police in each state in 2016") +
  theme(legend.title=element_blank()) 
```

![](readme-figs/unnamed-chunk-5-1.png)

But when we consider the differences in the number of people in each state, New Mexico and Alaska have the highest proportions of people killed by police.

``` r
ggplot(by_state, 
       aes(state = state_name, 
           fill = per_n_people)) +
  geom_statebins() +
  coord_equal() +
  scale_fill_viridis() +
  theme_statebins() +
  labs(title = "Number of people killed by police in each state in 2016,\nper 1,000,000 people")  +
  theme(legend.title=element_blank()) 
```

![](readme-figs/unnamed-chunk-6-1.png)

There does not appear to any obvious seasonality to police killings. June 2013 and July 2015 were especially bad months.

``` r
kbp2013_2017 %>% 
  filter(!is.na(event_month)) %>% 
  group_by(event_month, 
           event_year) %>% 
  tally() %>% 
  mutate(month_name = month.abb[event_month]) %>% 
  mutate(month_name = factor(month_name, levels = month.abb)) %>% 
ggplot() +
  geom_tile(aes(x = month_name, 
                y = event_year, 
                fill = n)) +
  scale_fill_viridis_c() +
  theme_minimal() +
  xlab("Month") +
  ylab("Year")
```

![](readme-figs/unnamed-chunk-7-1.png)

Have any states seen strong upward or downward trends over time? It's difficult to see if we plot them all, so let's search for states with strong trends.

``` r
n_over_time_by_state <- 
kbp2013_2017 %>% 
  group_by(state,
           event_year) %>% 
  tally() %>% 
  ungroup()

ggplot(n_over_time_by_state,
       aes(event_year,
           n,
           colour = state)) +
  geom_line() +
  geom_text(data = n_over_time_by_state[n_over_time_by_state$n > 50 & n_over_time_by_state$event_year == 2017,],
           aes(label = state,
               x = 2017.1, 
               y = n, 
               colour = state, 
               hjust = -.01)) +
  theme_minimal()  +
  ylab("Number of people killed by police") +
  xlab("Year") +
  theme(legend.position="none")
```

![](readme-figs/unnamed-chunk-8-1.png)

Let's compute a linear model for the number of killings and years for each state, then see which states have the most extreme trends. We can used a nested data frame to do this efficiently (based on the example in Grolemund and Wickham's book [*R for Data Science*](http://r4ds.had.co.nz/many-models.html)):

``` r
library(tidyr)
nested_df <- 
n_over_time_by_state %>% 
  filter(state != "") %>% 
  filter(!is.na(event_year)) %>% 
  nest(-state)
```

Which looks like this:

``` r
head(nested_df)
#> # A tibble: 6 x 2
#>   state data            
#>   <chr> <list>          
#> 1 AK    <tibble [5 x 2]>
#> 2 AL    <tibble [5 x 2]>
#> 3 AR    <tibble [5 x 2]>
#> 4 AZ    <tibble [5 x 2]>
#> 5 CA    <tibble [5 x 2]>
#> 6 CO    <tibble [5 x 2]>
```

And the *tibble* that we see in each row looks something like this:

``` r
nested_df$data[1]
#> [[1]]
#> # A tibble: 5 x 2
#>   event_year     n
#>        <dbl> <int>
#> 1       2013     2
#> 2       2014     2
#> 3       2015     6
#> 4       2016     8
#> 5       2017     8
```

Now let's compute linear models for all states:

``` r
library(purrr)
model_by_state <- 
nested_df %>% 
  mutate(model = map(data, ~lm(event_year ~ n, data = .)))
```

Let's only look at the states where the linear model has a p-value of less than 0.05 and the adjusted R<sup>2</sup> is &gt;0.7. This will limit us to states that have a statistically significant trend over time:

``` r
glance <- model_by_state %>% 
  mutate(glance = map(model, broom::glance)) %>% 
  unnest(glance, .drop = TRUE)

glance_sig <- 
  glance %>% 
  filter(adj.r.squared >= 0.7) %>% 
  filter(p.value <= 0.05) %>% 
  arrange(desc(adj.r.squared))

glance_sig
#> # A tibble: 7 x 12
#>   state r.squ~ adj.r.~ sigma stati~ p.value    df logLik   AIC   BIC devi~
#>   <chr>  <dbl>   <dbl> <dbl>  <dbl>   <dbl> <int>  <dbl> <dbl> <dbl> <dbl>
#> 1 AL     0.943   0.923 0.438   49.2 0.00595     2  -1.69  9.37  8.20 0.574
#> 2 CO     0.931   0.908 0.479   40.5 0.00785     2  -2.14 10.3   9.11 0.690
#> 3 WV     0.908   0.878 0.552   29.8 0.0121      2  -2.85 11.7  10.5  0.916
#> 4 AK     0.880   0.841 0.631   22.1 0.0182      2  -3.52 13.0  11.9  1.20 
#> 5 NM     0.849   0.799 0.709   16.9 0.0261      2  -4.10 14.2  13.0  1.51 
#> 6 WI     0.816   0.755 0.782   13.3 0.0355      2  -4.59 15.2  14.0  1.84 
#> 7 VA     0.785   0.713 0.847   10.9 0.0455      2  -4.99 16.0  14.8  2.15 
#> # ... with 1 more variable: df.residual <int>
```

And when we plot only these states with the signficant trends we see that they all indicate a trend of *increasing* deaths by police over time.

No state seems to have succeeded in reducing the total number of people killed by police over time.

``` r
n_over_time_by_state_sig <- 
n_over_time_by_state %>% 
  filter(state %in% glance_sig$state) 

ggplot(n_over_time_by_state_sig,
       aes(event_year,
           n,
           group = state,
           colour = state)) +
  geom_line() +
  geom_text(data = n_over_time_by_state_sig[ n_over_time_by_state_sig$event_year == 2017,],
           aes(label = state,
               x = 2017.1, 
               y = n, 
               colour = state, 
               hjust = -.01)) +
  theme_minimal() +
  ylab("Number of people killed by police") +
  xlab("Year")
```

![](readme-figs/unnamed-chunk-14-1.png)

Related work
------------

There are several Python scripts for scraping killedbypolice.net, for example by [markberger](https://github.com/markberger/police-killings-dataset) and [AceLewis](https://github.com/AceLewis/killedbypolice), and a Ruby script by [atom-morgan](https://github.com/atom-morgan/killed-by-police). We have not tried any of these, or used them here.

There are several other major projects with similar data:

-   [Mapping Police Violence](https://mappingpoliceviolence.org/) contains data on people killed by police since 2013. They aggregate data from [*Fatal Encounters*](http://www.fatalencounters.org/), killedbypolice.net, and the [U.S. Police Shootings Database](https://docs.google.com/spreadsheets/d/1cEGQ3eAFKpFBVq1k2mZIy5mBPxC6nBTJHzuSWtZQSVw/edit). They include data about whether the deceased was armed or not, and if a vehicle was involved, how it contributed to the death. The site features numerous visualisations exploring the data, especially the representation of Black people, and includes some interactive visualisations.
-   [*Fatal Encounters*](http://www.fatalencounters.org/) is a database of all deaths through police interaction in the United States since 1 Jan 2000. It is a public [Google spreadsheet](https://docs.google.com/spreadsheets/d/1dKmaV_JiWcG8XBoRgP8b4e9Eopkpgt7FL7nyspvzAsE/edit#gid=0). It contains variables not inlcuded in killedbypolice, such as street addresses and history of mental illness. Data are tweeted at [@fatalencounters](https://twitter.com/fatalencounters) and [@the\_decedents](https://twitter.com/the_decedents). They are still updating their data. It has been used by [Snopes.com](https://www.snopes.com/do-police-kill-more-whites-than-black-people/).
-   *The Washington Post* started compiling data on fatal shootings by U.S. police in 2015. Their data are available in CSV files at their [Washington Post GitHub repository](https://github.com/washingtonpost/data-police-shootings). They are still updating their data, and it includes some variables not collected by the killedbypolice, such as whether or not the deceased was armed, and what city the event occured in. These data have also been used by [Buzzfeed](https://github.com/BuzzFeedNews/2015-12-fatal-police-shootings).
-   *The Guardian* collected data on on fatal shootings by U.S. police in 2015-2016 for their project [The Counted](https://www.theguardian.com/us-news/ng-interactive/2015/jun/01/the-counted-police-killings-us-database). They are no longer updating this project. These data have been used by [FiveThirtyEight](https://github.com/fivethirtyeight/data/tree/master/police-killings) and [Buzzfeed](https://github.com/BuzzFeedNews/2015-12-fatal-police-shootings) and projects on GitHub, e.g. [flother](https://github.com/flother/thecounted), and [Kaggle](https://www.kaggle.com/the-guardian/the-counted).
-   [VICE News](https://news.vice.com/en_us/article/xwvv3a/shot-by-cops) examined both fatal and nonfatal incidents from 2010 through 2016. The data are on [GitHub](https://github.com/vicenews/shot-by-cops/) and in a [Google Sheet](https://docs.google.com/spreadsheets/d/1CaOQ7FUYsGFCHEqGzA2hlfj69sx3GE9GoJ40OcqI9KY/edit#gid=1271324584)
-   [Wikipedia](https://en.wikipedia.org/wiki/List_of_killings_by_law_enforcement_officers_in_the_United_States) has lists of killings by law enforcement offiers in the US, but the numbers are much lower than any of the other sources noted here.

Contributing
------------

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md
# killedbypolice 0.1.0

* Added a `NEWS.md` file to track changes to the package.


data-raw/read_data_from_webapges.R
#' Import the data from the webpages into R objects
#'
#------------------------------------------------------------------------------

#'
#' Here are the web pages that we downloaded on 1 Jan 2018, using Chrome Version 63.0.3239.84 (Official Build) (64-bit)
library(here)
webpages <- list.files(here("/data-raw"), pattern = ".html$", full.names = TRUE)

# read in the webapges into the R session
library(tidyverse)
all_pages <- map(webpages, ~read_table(.x))

#------------------------------------------------------------------------------


#' filter to keep only columns that have data in them, get rid of some HTML
x2 <- map(all_pages, ~.x %>%
            filter(grepl("<tr><td><center>", X1)))

library(glue)
x3 <- map(x2, ~.x %>%
            # split variables into cols
            separate(X1,
                     into = glue('V{1:10}'),
                     sep = "<td>",
                     remove = FALSE))

#------------------------------------------------------------------------------

#' convert list of data frames to one big data frame
x4 <- bind_rows(x3)

#------------------------------------------------------------------------------



# we want to separate rows on V2, V4, V5 (date, sex/race, name), but these are sometime
# not exactly equal when there are mulitple deceased on one event. So we need to
# tidy it a bit.

# how many <br> in each col?
x5 <- x4  %>%
          mutate(br_in_date = str_count(V2, "<br>")) %>%
          mutate(br_in_genderrace = str_count(V4, "<br>")) %>%
          mutate(br_in_nameage = str_count(V5, "<br>")) %>%
          mutate(max_br = pmax(br_in_date, br_in_genderrace, br_in_nameage)) %>%
            rowwise() %>%
          mutate(which_has_max = which.max(c(br_in_date,
                                             br_in_genderrace,
                                             br_in_nameage)))

# insert some <br> into cols if they don't have the max number for that row

x5$genderrace_new <- 0
x5$nameage_new <- 0
x5$date_new <- 0

for(i in seq_len(nrow(x5))){

  x5$genderrace_new[i] <-
    with(x5, ifelse((br_in_date[i] > br_in_genderrace[i]),
           paste(V4[i], rep("<br>.", (br_in_date[i] - br_in_genderrace[i]))),
           V4[i]))

  x5$genderrace_new[i] <-
    with(x5, ifelse(br_in_nameage[i] > str_count(genderrace_new[i], "<br>"),
                    paste(genderrace_new[i], rep("<br>.", (br_in_nameage[i] - br_in_genderrace[i]))),
                    genderrace_new[i]))

  x5$nameage_new[i] <-
    with(x5, ifelse((br_in_date[i] > br_in_nameage[i]),
           paste(V5[i], rep("<br>.", (br_in_date[i] - br_in_nameage[i]))),
           V5[i]))

  x5$date_new[i] <-
    with(x5, ifelse((br_in_genderrace[i] > br_in_date[i]),
                    paste(V2[i], rep(glue("<br>.{V2[i]}"), (br_in_genderrace[i] - br_in_date[i]))),
                    V2[i]))

  x5$date_new[i] <-
    with(x5, ifelse(br_in_nameage[i] > str_count(date_new[i], "<br>"),
                    paste(date_new[i], rep(glue("<br>.{date_new[i]}"), (br_in_nameage[i] - br_in_date[i]))),
                    date_new[i]))


}

# check that we have the same number of <br> in those three cols
# x6 <- x5  %>%
#   mutate(br_in_date2 = str_count(date_new, "<br>")) %>%
#   mutate(br_in_genderrace2 = str_count(genderrace_new, "<br>")) %>%
#   mutate(br_in_nameage2 = str_count(nameage_new, "<br>")) %>%
#   mutate(max_br2 = pmax(br_in_date2, br_in_genderrace2, br_in_nameage2))
#------------------------------------------------------------------------------


#' Continue cleaning, format dates, get rid of white space, before we separate rows
x6 <- x5 %>%
  mutate_all(funs(trimws)) %>%
  mutate(date = str_replace_all(date_new,
                                "<center>\\([0-9]*\\)|</center></td>|[:punct:]",
                                "")) %>%
  mutate_all(funs(trimws)) %>%
  mutate(date_format =  as.Date(date, "%B %d %Y"))

#------------------------------------------------------------------------------


#' Now that we've made sure we have an equal number of <br> in each col,
#' we can separate rows on <br> to get multiple deceased in one event
#' in one row each
x7 <- x6 %>%
      separate_rows(c(date_new, genderrace_new, nameage_new),
                     sep = "<br>")

#------------------------------------------------------------------------------


#'  remove URLs in name col, split deceased into name and age
x8 <- x7 %>%
            mutate(deceased = gsub('http\\S+\\s*', "", nameage_new)) %>%
            mutate(deceased = gsub('<a href="|</a>|target="new"|</td>', "", deceased)) %>%
            mutate(deceased = str_replace_all(deceased, "<|>", "")) %>%
            # get the name and age in their own cols
            separate(deceased,
                     into = c('deceased_name', 'deceased_age'),
                     sep = ",(?=[^,]+$)",  # split on the last comma
                     remove = FALSE)  %>%
            mutate(deceased_age = as.numeric(trimws(deceased_age)))

#------------------------------------------------------------------------------


#' split gender and race col into separate cols
x9 <-  x8 %>%
            mutate(genderrace_new = str_replace_all(genderrace_new, "</td>", "")) %>%
            separate(genderrace_new,
                     into = c('gender',
                              'race'),
                     sep = "/")

#------------------------------------------------------------------------------


#' separate out multiple methods. We have to be careful here because we can see
#' multiple methods where there are multiple deceased on one event, and where
#' there is one deceased who was killed in multiple ways. Clean out the HTML
#' first...

x10 <- x9 %>%
            mutate(V6 = str_replace_all(V6, '<font size="2">|</font></td>|<center>|</center>|<font color="white">\\.</font>', "")) %>%
            mutate(method = trimws(V6))

#' Then if there are two methods, and the second is different from the first,
#' then we assume that it's one deceased with multiple methods, and we want
#' to keep those. We don't want to keep multiple same methods, not informative
#'
# unique(x10$method)
method_simplify <- function(x){
  t2 <- str_extract_all(x, "[A-Z]")
  t3 <- map_int(t2, ~length(unique(.x)))
  t4 <- map2(t2, t3, ~.x[1:.y])
  t5 <- map_chr(t4, ~paste(.x, collapse = "<br>"))
  return(t5)
}

x10$method <- method_simplify(x10$method)

x11 <-  x10 %>%
        separate(method,
                 into = glue('method_{1:5}'),
                 sep = "<br>") %>%
  mutate_all(funs(trimws))
# unique(x11$method_3)

#------------------------------------------------------------------------------


#' tidy state column
x12 <- x11 %>%
            mutate(State = str_replace_all(V3, "</td>", "")) %>%
            mutate(State = trimws(State))

#------------------------------------------------------------------------------


#' get only columns of interest & ensure type is ok
x13 <- x12 %>%
  select("date_format",
         "State",
         "deceased_name",
         "deceased_age",
         "gender",
         "race",
          "method_1",
          "method_2",
          "method_3",
          "method_4",
          "method_5"
          ) %>%
  mutate(deceased_age = as.numeric(deceased_age),
         date_format = as.Date(date_format))

# rename a few variables, tidy date
library(lubridate)
x14 <- x13 %>%
  rename(event_date = date_format,
         state = State,
         name = deceased_name,
         age = deceased_age,
         race_ethnicity = race) %>%
  mutate(event_year = year(event_date),
         event_month = month(event_date),
         event_day = day(event_date)) %>%
  select(-event_date)

# any stray HTML still in there?
# look for it
x14 %>%
   filter_all(any_vars(grepl("<|>", .))) %>%
   View

# clear it out
x15 <- x14 %>%
  mutate_all(funs(ifelse(grepl("<|>", .), "", .)))

x15$gender <- str_replace_all(x15$gender, "\\.", "")

#' Now we have the data frame for 2013-2017
#' Let's save to the package
kbp2013_2017 <- x15
devtools::use_data(kbp2013_2017, overwrite = TRUE)

#' and document


#------------------------------------------------------------------------------

#------------------------------------------------------------------------------

#' What does it look like? Here's a snippet of just a few columns:
x13 %>%
  select(date_format,
         State,
         deceased_name,
         deceased_age) %>%
  head(n = 10) %>%
  knitr::kable()

#' ## Some quick plots

ggplot(x13,
       aes(deceased_age)) +
  geom_histogram() +
  theme_bw()

#' age and gender
age_and_gender <-
  x13 %>%
  select(deceased_age, gender) %>%
  filter(!is.na(deceased_age)) %>%
  mutate(age_category=cut(deceased_age,
                          breaks = seq(0,max(deceased_age), 5))) %>%
  filter(!is.na(age_category)) %>%
  group_by(age_category) %>%
  summarise(prop_female = sum(gender == 'F') / n())

ggplot(age_and_gender,
       aes(age_category,
           prop_female)) +
  geom_point(size = 3) +
  ylab("Proportion of deaths that are Female") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  ggtitle("Female deaths by police, by age group",
          subtitle = glue("Data from kpb"))

#' race
race <-
  x13 %>%
  select(deceased_age, race) %>%
  filter(!is.na(race)) %>%
  filter(!is.na(deceased_age)) %>%
  mutate(race = str_sub(race, 1, 1)) %>%
  filter(race %in% c("I", "B", "W", "L", "A", "P", "O"))

# get n to use in tick labels
race_n <-
  race %>%
  group_by(race) %>%
  tally() %>%
  mutate(race_label = glue('{race}\n(n = {n})')) %>%
  right_join(race)

library(ggbeeswarm)
ggplot(race_n,
       aes(reorder(race_label,
                   -deceased_age),
           deceased_age)) +
  geom_quasirandom(alpha = 0.1) +
  xlab("Race (as classified by killedbypolice.net)") +
  theme_bw()  +
  ggtitle("Deaths by police, by age and race of deceased",
          subtitle = glue("Data from kpb"))

#' method of death
method_of_death <-
  x13 %>%
  mutate(method_long = case_when(
    method_1 == "G" ~ "Gun",
    method_1 == "T" ~ "Taser",
    method_1 == "R" ~ "Restraint/\nPhysical Force",
    method_1 == "C" ~ "Chemical",
    method_1 == "V" ~ "Vehicle",
    method_1 == "O" ~ "Other"))

method_of_death_df <-
  method_of_death %>%
  group_by(method_long) %>%
  tally() %>%
  mutate(prop = n / sum(n) ) %>%
  filter(!is.na(method_long))


ggplot(method_of_death_df,
       aes(reorder(method_long, -prop),
           prop)) +
  geom_col() +
  xlab("") +
  ylab("Proportion of all\ndeaths by police") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

#' Let's convert n to the proportion of people in each state

#' age distribution by state
#' https://hafen.github.io/geofacet/
library("geofacet")

ggplot(x13,
       aes(deceased_age)) +
  geom_histogram() +
  theme_bw() +
  facet_geo(~ State)

#'  hexbin state map
#'  based on https://rud.is/b/2015/05/15/u-s-drought-monitoring-with-hexbin-state-maps-in-r/
#'  get state population data
tmp <- rio::import("https://www2.census.gov/programs-surveys/popest/tables/2010-2016/state/totals/nst-est2016-01.xlsx",
                   skip = 3)

# save it locally
write.csv(tmp, paste0(here::here("data-raw"),"/nst-est2016-01.csv"))

# clean up state names to abbreviations

state_abb <- data_frame(state_name = state.name,
                        state_abb = state.abb)
state_abb <- rbind(state_abb,
                   data_frame(
                     state_name = "District of Columbia",
                     state_abb = "DC"))
state_pops <-
  tmp %>%
  mutate(state_name = gsub("\\.", "", X__1)) %>%
  left_join(state_abb)

x14 <-
  x13 %>%
  group_by(State) %>%
  tally() %>%
  left_join(state_pops, by = c("State" = "state_abb")) %>%
  mutate(prop =  n / Census * 100000) # per 100k people

library(rgdal)
library(rgeos)

# get map from
download.file("https://gist.githubusercontent.com/hrbrmstr/51f961198f65509ad863/raw/219173f69979f663aa9192fbe3e115ebd357ca9f/us_states_hexgrid.geojson", destfile=paste0(here::here("data-raw"),"/us_states_hexgrid.geojson"))
us <- readOGR(paste0(here::here("data-raw"),"/us_states_hexgrid.geojson"), "OGRGeoJSON")
centers <- cbind.data.frame(data.frame(gCentroid(us, byid=TRUE), id=us@data$iso3166_2))
us_map <- fortify(us, region="iso3166_2")

gg <- ggplot()

gg <- gg + geom_map(
  data = us_map,
  map = us_map,
  aes(x = long,
      y = lat,
      map_id = id),
  color = "white",
  size = 0.5
)

gg <- gg + geom_map(data = x7,
                    map = us_map,
                    aes(fill = prop,
                        map_id = V3))

gg <- gg + geom_map(
  data = x7,
  map = us_map,
  aes(map_id = V3),
  fill = "#ffffff",
  alpha = 0,
  color = "white",
  show.legend = FALSE
)

gg <- gg + geom_text(data=centers, aes(label=id, x=x, y=y), color="white", size=4)

gg <- gg + coord_map() +
  scale_fill_viridis_c() +
  theme_bw()  +
  xlab("") +
  ylab("")

gg <- gg + theme(panel.border=element_blank())
gg <- gg + theme(panel.spacing=unit(3, "lines"))
gg <- gg + theme(panel.grid=element_blank())
gg <- gg + theme(axis.ticks=element_blank())
gg <- gg + theme(axis.text=element_blank())
gg <- gg + theme(strip.background=element_blank())
gg <- gg + theme(strip.text=element_text(face="bold", hjust=0, size=14))
gg <- gg + ggtitle(glue('Killed by police, per 100k people, {min(years)}-2017'),
                   subtitle = "Data from http://killedbypolice.net/")

gg
#' This seems right, cf. https://mappingpoliceviolence.org/states/
#'
#' interactive map of raw counts by state over time
#'
state_pops_long <-
  state_pops %>%
  select(state_abb, one_of(as.character(years))) %>%
  gather(year, population, -state_abb) %>%
  filter(!is.na(state_abb))


x8 <-
  x6 %>%
  mutate(month = format(date_format, "%m"),
         year = format(date_format, "%Y"),
         month_year = glue('{month}/{year}'),
         state_abb = V3) %>%
  group_by(month, year, month_year, state_abb) %>%
  tally() %>%
  arrange(year, month) %>%
  left_join(state_pops_long)

#'  get an ordered factor to make the x-axis plot in order
x8$month_year <- factor(x8$month_year, levels = unique(x8$month_year))

gg1 <-
  ggplot(x8,
         aes(month_year,
             n)) +
  geom_line(aes(group = state_abb,
                colour = state_abb)) +
  scale_color_viridis_d() +
  # scale_y_log10() +
  theme_bw()  +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))  +
  ggtitle(glue('Raw counts of people killed by police, {min(years)}-2017'),
          subtitle = "Data from http://killedbypolice.net/")

gg1

# get interactive plot
plotly::ggplotly(gg1)


#' ## No need to read below here
#' This is just drafts of the code that I used above

#------------------------------------------------------------------------------
# No need to look here, it's just what I did to get the 2017 data only
# for one webpage, the KBP 2017 data
# get HTML page as raw HTML
tmp1 <- readr::read_table(url)
# rename the column for ease of use
names(tmp1) <- "V1"


# filter to keep only columns that have data in them, get rid of some HTML
tmp2 <- tmp1 %>%
  mutate(V99 = gsub('<center>|<CENTER>|<font size=2>|target=new', "", V1)) %>%
  filter(grepl("<TR><TD>\\(", V99)) %>%
  select(-V1)


tmp3 <- tmp2 %>%
  # split variables into cols
  separate(V99,
           into = glue('V{1:9}'),
           sep = "<td>|<TD>",
           remove = FALSE) %>% # keep it so we can check on things later
  separate(V2,
           into = c('number', 'date',
                    'number2', 'date2'),
           sep = "\\)|<br>|\\)|<BR>")

# format some col classes, numeric, date, get rid of white space
tmp4 <-
  tmp3 %>%
  mutate_all(funs(trimws)) %>%
  mutate(number = as.numeric(gsub("\\(", "", number))) %>%
  mutate(date_format =  as.Date(date, "%B %d, %Y"))

# remove URLs in name col, split mulitple desceased into multiple cols
tmp5 <-
  tmp4 %>%
  mutate(deceased = gsub('http\\S+\\s*', "", V5)) %>%
  mutate(deceased = gsub('<a href=">|</a>', "", deceased)) %>%
  # make separate row where there are two deceased in one row
  separate_rows(deceased, sep = "<br>") %>%
  # get the name and age in their own cols
  separate(deceased,
           into = c('deceased_name', 'deceased_age'),
           sep = ",(?=[^,]+$)",  # split on the last comma
           remove = FALSE)  %>%
  mutate(deceased_age = as.numeric(trimws(deceased_age))) %>%
  separate(V4,
           into = c('gender', 'race'),
           sep = "/")


# How many are we missing?
max_number_on_website <- max(tmp5$number)
number_that_we_have <- nrow(tmp5)
difference <- max_number_on_website - number_that_we_have
difference


# is this from your group? https://rpubs.com/johnbradford/policeShootDataCreation




